src/applications/bmqstoragetool/m_bmqstoragetool_parameters.cpp

```diff
@@ -251,11 +253,143 @@ bool Parameters::partiallyConfirmed() const
+        new (*allocator) mqbmock::LogIdGenerator("bmq_csl_", allocator),
+        allocator);
+    bsl::shared_ptr<mqbsi::LogFactory> logFactory(
+        new (*allocator) mqbsl::MemoryMappedOnDiskLogFactory(allocator),
+        allocator);
+    auto fileSize = bdls::FilesystemUtil::getFileSize(d_cslFile.c_str());
+    bsl::string pattern(allocator);
+    bsl::string location(allocator);
+    BSLS_ASSERT(bdls::PathUtil::getBasename(&pattern, d_cslFile) == 0);
+    BSLS_ASSERT(bdls::PathUtil::getDirname(&location, d_cslFile) == 0);
+    ledgerConfig.setLocation(location)
+        .setPattern(pattern)
+        .setMaxLogSize(
+            fileSize)  // TODO: what size should be? Assume size of Cls file.
+        .setReserveOnDisk(false)
+        .setPrefaultPages(false)
+        .setLogIdGenerator(logIdGenerator)
+        .setLogFactory(logFactory)
+        .setExtractLogIdCallback(&mqbc::ClusterStateLedgerUtil::extractLogId)
+        .setValidateLogCallback(mqbc::ClusterStateLedgerUtil::validateLog);
+
+    // Create and open the ledger
+    mqbsl::Ledger ledger(ledgerConfig, allocator);
+    BSLS_ASSERT(ledger.open(mqbsi::Ledger::e_READ_ONLY) == 0);
+    // Set guard to close the ledger
+    auto closeLedger = [](mqbsl::Ledger* ledger) {
+        BSLS_ASSERT(ledger->close() == 0);
+    };
+    bdlb::ScopeExitAny guard(bdlf::BindUtil::bind(closeLedger, &ledger));
+
+    // Iterate through each record in the ledger to find the last snapshot
+    // record
+    mqbc::IncoreClusterStateLedgerIterator cslIt(&ledger);
+    mqbc::IncoreClusterStateLedgerIterator lastSnapshotIt(&ledger);
+    while (true) {
+        int rc = cslIt.next();
+        if (rc == 1) {
+            if (!lastSnapshotIt.isValid()) {
+                ss << "No Snapshot found in csl file." << bsl::endl;
+                return false;  // RETURN
+            }
+            break;
+        }
+        else if (rc < 0) {
+            ss << "Csl iterator error: " << rc << bsl::endl;
+            return false;  // RETURN
+        }
+
+        if (cslIt.header().recordType() ==
+            mqbc::ClusterStateRecordType::Enum::e_SNAPSHOT) {
+            // Save snapshot iterator
+            lastSnapshotIt = cslIt;
+        }
+    }
+
+    // Process last snapshot
+    bmqp_ctrlmsg::ClusterMessage clusterMessage;
+    lastSnapshotIt.loadClusterMessage(&clusterMessage);
+    BSLS_ASSERT(
+        clusterMessage.choice().selectionId() ==
+        bmqp_ctrlmsg::ClusterMessageChoice::SELECTION_ID_LEADER_ADVISORY);
+
+    // Get queue info from snapshot (leaderAdvisory) record
+    auto leaderAdvisory = clusterMessage.choice().leaderAdvisory();
+    auto queuesInfo     = leaderAdvisory.queues();
+
+    // Helper lambda to fill queue maps
+    auto& queueKeyToInfoMap = d_queueInfo.queueKeyToInfoMap();
+    auto& queueUriToKeyMap  = d_queueInfo.queueUriToKeyMap();
+    auto  fillMap           = [&](const bmqp_ctrlmsg::QueueInfo& queueInfo) {
```

Comment:
Seems it should be a method of QueueInfo class for adding incapsulating the manipulations with its internal hashmaps. Moreover, let's make QueueInfo class a bit smarter. Suggest adding some methods for frequently used opperations (like searching by uri or by Key) instead of giving access to its private members:
void QueueInfo::insert(const bmqp_ctrlmsg::QueueInfo& queueInfo)
bmqp_ctrlmsg::QueueInfo QueueInfo::findKey(const mqbu::StorageKey& key) const
mqbu::StorageKey QueueInfo::findUri(const bsl::string& uri) const

Reply:
Added suggested methods and moved logic there. Removed access to private members.